File: DevEnvironmentCLion/CANMessageGeneratorCommand/core/CANMessageGeneratorCommand.cpp
#include "PreCompiled.h"
#include "CANMessageGeneratorCommand.h"
#include "CANMessageGeneratorCommandParameters.h"
#include "CANMessageRuleGenerator.h"
#include "JitterMessageHandler.h"
#include "Threads.h"

using namespace controllerareanetwork::messagegeneratorcommand;

const std::string CANMessageGeneratorCommand::messagesDeviationText = "Controller Area Network Message Generator Summary:";
const std::string CANMessageGeneratorCommand::messageNameText = "messageName:";

/**
 * Creates a new command.
 */
command::ICommandPtr CANMessageGeneratorCommand::make()
{
    return std::make_shared<CANMessageGeneratorCommand>();
}

/**
 * The CANMessageGeneratorCommand constructor.
 */
CANMessageGeneratorCommand::CANMessageGeneratorCommand() :
    m_messageRuleGenerator(std::make_unique<CANMessageRuleGenerator>()),
    m_messageHandler(std::make_unique<JitterMessageHandler>())
{
}

/**
 * The CANMessageGeneratorCommand destructor.
 */
CANMessageGeneratorCommand::~CANMessageGeneratorCommand()
{
}

/**
 * Runs the logic of a command.
 */
void CANMessageGeneratorCommand::run()
{
    initialize();

    generateMessages(
        m_parameters->getDriveSessionDuration(),
        m_parameters->getMessagesFrequencies());

    getMessageWriter().writeInformationalMessage("\n");

    displayMessagesJitterInformation(
        m_parameters->getMessagesFrequencies(),
        getMessageWriter());
}

/**
 * Generates the specified messages.
 */
void CANMessageGeneratorCommand::generateMessages(
    const base::Duration& driveSessionDuration,
    const std::vector<std::pair<std::string, size_t>>& messages)
{
    registerMessagesWithJitter(messages);

    base::DateTimePtr startTime = base::DateTime::now();

    bool processSession = true;

    while (processSession)
    {
        for (const std::pair<std::string, size_t>& messageFrequency : messages) {
            processSession = sessionActive(*startTime, driveSessionDuration);

            if (!processSession) {
                break;
            }

            const std::string& messageName = messageFrequency.first;
            size_t frequency = messageFrequency.second;

            base::DurationPtr timeout = base::Duration::fromSeconds(1);

            processMessages(
                messageName,
                frequency,
                *timeout,
                getMessageWriter());
        }
    }
}

/**
 * Processes the messages.
 */
bool CANMessageGeneratorCommand::processMessages(
    const std::string& messageName,
    size_t numberOfMessages,
    const base::Duration& timeout,
    base::IMessageWriter& messageWriter)
{
    size_t messageIndex = 0;

    base::DateTimePtr startTime = base::DateTime::now();

    size_t delayInMilliseconds = 1000 / numberOfMessages;

    while (messageIndex < numberOfMessages && sessionActive(*startTime, timeout))
    {
        processMessage(messageName, messageWriter);
        ++messageIndex;

        base::Threads::sleep(delayInMilliseconds);
    }

    return true;
}

/**
 * Processes the message.
 */
void CANMessageGeneratorCommand::processMessage(
    const std::string& messageName,
    base::IMessageWriter& messageWriter)
{
    base::DateTimePtr sendingTime = base::DateTime::now();
    m_messageHandler->setMessageSendingTime(messageName, sendingTime);

    CANMessageRuleGenerator ruleGenerator;
    ICANMessageRulePtr rule = ruleGenerator.generate(messageName);

    messageWriter.writeInformationalMessage(rule->toString());

    for (ICANSignalRulePtr signalRule : rule->getSignalRules())
    {
        messageWriter.writeInformationalMessage("    " + signalRule->toString());
    }

    m_messageHandler->setMessageReceivingTime(messageName, base::DateTime::now());
}

/**
 * Displays the Jitter information of the messages.
 */
void CANMessageGeneratorCommand::displayMessagesJitterInformation(
    const std::vector<std::pair<std::string, size_t>>& messages,
    base::IMessageWriter& messageWriter)
{
    messageWriter.writeInformationalMessage(messagesDeviationText);

    for (const std::pair<std::string, size_t> message : messages)
    {
        displayMessageJitterInformation(message.first, messageWriter);
    }
}

/**
 * Displays the Jitter information of the message.
 */
void CANMessageGeneratorCommand::displayMessageJitterInformation(
    const std::string& messageName,
    base::IMessageWriter& messageWriter)
{
    IJitterMessageDataPtr messageData = m_messageHandler->getMessageData(messageName);

    std::stringstream informationStream;
    informationStream
        << messageNameText << " " << messageName << ", " << messageData->toString();

    messageWriter.writeInformationalMessage(informationStream.str());
}

/**
 * Registers the messages with the Jitter.
 */
void CANMessageGeneratorCommand::registerMessagesWithJitter(const std::vector<std::pair<std::string, size_t>>& messages)
{
    for (const std::pair<std::string, size_t>& message : messages)
    {
        const std::string& messageName = message.first;
        size_t frequencyInHz = message.second;

        size_t transmittingTimeInMilliseconds = 1000 / frequencyInHz;
        base::DurationPtr transmittingTime = base::Duration::fromMilliseconds(transmittingTimeInMilliseconds);

        m_messageHandler->registerMessage(messageName, transmittingTime);
    }
}

/**
 * Checks whether the session is active.
 */
bool CANMessageGeneratorCommand::sessionActive(const base::DateTime& startTime, const base::Duration& timeout)
{
    base::DateTimePtr currTime = base::DateTime::now();

    base::DurationPtr currDuration = base::Duration::between(startTime, *currTime);

    return *currDuration < timeout;
}

/**
 * Initializes the command.
 */
void CANMessageGeneratorCommand::initialize()
{
    const command::ICommandParameters& parameters = getParameters();
    m_parameters = std::make_unique<CANMessageGeneratorCommandParameters>(parameters);
}

File: DevEnvironment/CANMessageGeneratorCommand/src/main/java/canmessagegeneratorcommand/core/CANMessageGeneratorCommand.java
package canmessagegeneratorcommand.core;

import base.core.Threads;
import base.interfaces.IPair;
import canmessagegenerator.core.CANMessageRuleGenerator;
import canmessagegenerator.core.JitterMessageHandler;
import canmessagegenerator.interfaces.ICANMessageRule;
import canmessagegenerator.interfaces.ICANRuleGenerator;
import canmessagegenerator.interfaces.ICANSignalRule;
import canmessagegenerator.interfaces.IJitterMessageData;
import canmessagegenerator.interfaces.IJitterMessageHandler;
import canmessagegeneratorcommand.interfaces.ICANMessageGeneratorCommandParameters;
import command.core.AbstractCommand;
import java.time.Duration;
import java.time.Instant;
import java.util.List;
import base.interfaces.IMessageWriter;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * The CANMessageGeneratorCommand class implements a command for a controller area network.
 */
public final class CANMessageGeneratorCommand extends AbstractCommand {
    private static final String messagesDeviationText = "Controller Area Network Message Generator Summary:";
    private static final String messageNameText = "messageName:";

    private ICANMessageGeneratorCommandParameters parameters;
    private final IJitterMessageHandler messageHandler = new JitterMessageHandler();

    private final Logger log = LoggerFactory.getLogger(this.getClass());

    /**
     * The CANMessageGeneratorCommand constructor.
     */
    public CANMessageGeneratorCommand() {
    }

    /**
     * Runs the command.
     */
    @Override
    public void run() {
        this.initialize();

        this.generateMessages(
            this.parameters.getDriveSessionDuration(),
            this.parameters.getMessagesFrequencies());

        this.getMessageWriter().writeInformationalMessage("\n");

        this.displayMessagesJitterInformation(
            this.parameters.getMessagesFrequencies(),
            this.getMessageWriter());
    }

    /**
     * Generates the specified messages.
     */
    private void generateMessages(
        Duration driveSessionDuration,
        List<IPair<String, Integer>> messages) {

        this.registerMessagesWithJitter(messages);

        Instant startTime = Instant.now();

        boolean processSession = true;

        while (processSession) {
            for (IPair<String, Integer> messageFrequency : messages) {
                processSession = sessionActive(startTime, driveSessionDuration);

                if (!processSession) {
                    break;
                }

                String messageName = messageFrequency.first();
                int frequency = messageFrequency.second();

                Duration timeout = Duration.ofSeconds(1);

                processMessages(messageName, frequency, timeout, this.getMessageWriter());
            }
        }
    }

    /**
     * Processes the messages.
     */
    private boolean processMessages(
        String messageName,
        int numberOfMessages,
        Duration timeout,
        IMessageWriter messageWriter) {

        int messageIndex = 0;

        Instant startTime = Instant.now();

        int delayInMilliseconds = 1000 / numberOfMessages;

        while (messageIndex < numberOfMessages && sessionActive(startTime, timeout)) {
            processMessage(messageName, messageWriter);
            ++messageIndex;

            Threads.sleep(delayInMilliseconds);
        }

        return true;
    }

    /**
     * Processes the message.
     */
    private void processMessage(
        String messageName,
        IMessageWriter messageWriter) {

        this.messageHandler.setMessageSendingTime(messageName, Instant.now());

        ICANRuleGenerator<ICANMessageRule> ruleGenerator = new CANMessageRuleGenerator();
        ICANMessageRule rule = ruleGenerator.generate(messageName);

        messageWriter.writeInformationalMessage(rule.toString());

        for (ICANSignalRule signalRule : rule.getSignalRules()) {
            messageWriter.writeInformationalMessage("    " + signalRule);
        }

        this.messageHandler.setMessageReceivingTime(messageName, Instant.now());
    }

    /**
     * Displays the Jitter information of the messages.
     */
    private void displayMessagesJitterInformation(
        List<IPair<String, Integer>> messages,
        IMessageWriter messageWriter) {

        messageWriter.writeInformationalMessage(messagesDeviationText);

        for (IPair<String, Integer> message : messages) {
            this.displayMessageJitterInformation(message.first(), messageWriter);
        }
    }

    /**
     * Displays the Jitter information of the message.
     */
    private void displayMessageJitterInformation(
        String messageName,
        IMessageWriter messageWriter) {

        IJitterMessageData messageData = this.messageHandler.getMessageData(messageName);

        StringBuilder information = new StringBuilder();
        information
            .append(messageNameText).append(" ").append(messageName)
            .append(", ")
            .append(messageData);

        messageWriter.writeInformationalMessage(information.toString());
    }

    /**
     * Registers the messages with the Jitter.
     */
    private void registerMessagesWithJitter(List<IPair<String, Integer>> messages) {
        for (IPair<String, Integer> message : messages) {
            String messageName = message.first();
            int frequencyInHz = message.second();

            long transmittingTimeInMilliseconds = 1000 / frequencyInHz;
            Duration transmittingTime = Duration.ofMillis(transmittingTimeInMilliseconds);

            this.messageHandler.registerMessage(messageName, transmittingTime);
        }
    }

    /**
     * Checks whether the session is active.
     */
    private static boolean sessionActive(Instant startTime, Duration timeout) {
        Instant currTime = Instant.now();
        Duration currDuration = Duration.between(startTime, currTime);

        return currDuration.compareTo(timeout) < 0;
    }

    /**
     * Initializes the command.
     */
    private void initialize() {
        this.parameters = new CANMessageGeneratorCommandParameters(this.getParameters());
    }
}
